File: plugin/src/knowledge/providers/musicbrainz.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const MUSICBRAINZ_API = "https://musicbrainz.org/ws/2";
const USER_AGENT = "VaultTree/0.1.0 (https://github.com/drooamor/vault-tree)";

interface MusicBrainzArtistResult {
  artists?: Array<{
    id: string;
    name: string;
    type?: string;
    country?: string;
    "life-span"?: {
      begin?: string;
      end?: string;
      ended?: boolean;
    };
    disambiguation?: string;
    tags?: Array<{ name: string; count: number }>;
  }>;
}

interface MusicBrainzReleaseResult {
  releases?: Array<{
    id: string;
    title: string;
    date?: string;
    country?: string;
    "artist-credit"?: Array<{
      artist: { name: string };
    }>;
    "release-group"?: {
      "primary-type"?: string;
    };
  }>;
}

interface MusicBrainzRecordingResult {
  recordings?: Array<{
    id: string;
    title: string;
    length?: number;
    "artist-credit"?: Array<{
      artist: { name: string };
    }>;
    releases?: Array<{
      title: string;
      date?: string;
    }>;
  }>;
}

export class MusicBrainzProvider implements KnowledgeProvider {
  readonly name = "musicbrainz";

  async isAvailable(): Promise<boolean> {
    try {
      const response = await requestUrl({
        url: `${MUSICBRAINZ_API}/artist?query=test&limit=1&fmt=json`,
        method: "GET",
        headers: { "User-Agent": USER_AGENT },
        throw: false,
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      const entries: KnowledgeEntry[] = [];

      // Search artists first
      const artistResults = await this.searchArtists(query, maxResults);
      entries.push(...artistResults);

      // If few results, search releases (albums)
      if (entries.length < maxResults) {
        const releaseResults = await this.searchReleases(query, maxResults - entries.length);
        entries.push(...releaseResults);
      }

      // If still few results, search recordings (songs)
      if (entries.length < maxResults) {
        const recordingResults = await this.searchRecordings(query, maxResults - entries.length);
        entries.push(...recordingResults);
      }

      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `MusicBrainz lookup failed: ${message}`,
      };
    }
  }

  private async searchArtists(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${MUSICBRAINZ_API}/artist?query=${encodeURIComponent(query)}&limit=${limit}&fmt=json`,
      method: "GET",
      headers: { "User-Agent": USER_AGENT },
      throw: false,
    });

    if (response.status !== 200) {
      return [];
    }

    const data = response.json as MusicBrainzArtistResult;

    if (!data.artists || !Array.isArray(data.artists)) {
      return [];
    }

    return data.artists.map((artist) => {
      const lifespan = artist["life-span"];
      const years =
        lifespan?.begin || lifespan?.end
          ? ` (${lifespan.begin ?? "?"} - ${lifespan.ended ? lifespan.end ?? "" : "present"})`
          : "";

      const type = artist.type ? `${artist.type}` : "Artist";
      const country = artist.country ? `, ${artist.country}` : "";
      const disambiguation = artist.disambiguation ? ` - ${artist.disambiguation}` : "";

      const topTags = artist.tags
        ?.sort((a, b) => b.count - a.count)
        .slice(0, 3)
        .map((t) => t.name)
        .join(", ");

      return {
        title: artist.name,
        summary: `${type}${country}${years}${disambiguation}${topTags ? `. Genres: ${topTags}` : ""}`,
        url: `https://musicbrainz.org/artist/${artist.id}`,
        source: "musicbrainz",
        metadata: {
          type: "artist",
          artistType: artist.type,
          country: artist.country,
          beginDate: lifespan?.begin,
          endDate: lifespan?.end,
          tags: artist.tags?.map((t) => t.name),
        },
      };
    });
  }

  private async searchReleases(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${MUSICBRAINZ_API}/release?query=${encodeURIComponent(query)}&limit=${limit}&fmt=json`,
      method: "GET",
      headers: { "User-Agent": USER_AGENT },
      throw: false,
    });

    if (response.status !== 200) {
      return [];
    }

    const data = response.json as MusicBrainzReleaseResult;

    if (!data.releases || !Array.isArray(data.releases)) {
      return [];
    }

    return data.releases.map((release) => {
      const artists =
        release["artist-credit"]?.map((ac) => ac.artist.name).join(", ") ?? "Unknown artist";
      const year = release.date ? ` (${release.date.slice(0, 4)})` : "";
      const type = release["release-group"]?.["primary-type"] ?? "Release";

      return {
        title: release.title,
        summary: `${type} by ${artists}${year}`,
        url: `https://musicbrainz.org/release/${release.id}`,
        source: "musicbrainz",
        metadata: {
          type: "release",
          releaseType: release["release-group"]?.["primary-type"],
          artists: release["artist-credit"]?.map((ac) => ac.artist.name),
          date: release.date,
          country: release.country,
        },
      };
    });
  }

  private async searchRecordings(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${MUSICBRAINZ_API}/recording?query=${encodeURIComponent(query)}&limit=${limit}&fmt=json`,
      method: "GET",
      headers: { "User-Agent": USER_AGENT },
      throw: false,
    });

    if (response.status !== 200) {
      return [];
    }

    const data = response.json as MusicBrainzRecordingResult;

    if (!data.recordings || !Array.isArray(data.recordings)) {
      return [];
    }

    return data.recordings.map((recording) => {
      const artists =
        recording["artist-credit"]?.map((ac) => ac.artist.name).join(", ") ?? "Unknown artist";

      const duration = recording.length
        ? ` [${formatDuration(recording.length)}]`
        : "";

      const album = recording.releases?.[0]?.title
        ? ` from "${recording.releases[0].title}"`
        : "";

      return {
        title: recording.title,
        summary: `Recording by ${artists}${duration}${album}`,
        url: `https://musicbrainz.org/recording/${recording.id}`,
        source: "musicbrainz",
        metadata: {
          type: "recording",
          artists: recording["artist-credit"]?.map((ac) => ac.artist.name),
          duration: recording.length,
          releases: recording.releases?.map((r) => r.title),
        },
      };
    });
  }
}

function formatDuration(ms: number): string {
  const seconds = Math.floor(ms / 1000);
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = seconds % 60;
  return `${minutes}:${remainingSeconds.toString().padStart(2, "0")}`;
}

File: crates/mcp/src/knowledge/musicbrainz.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const MUSICBRAINZ_API: &str = "https://musicbrainz.org/ws/2";

pub struct MusicBrainzProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
struct ArtistSearchResponse {
    artists: Option<Vec<Artist>>,
}

#[derive(Debug, Deserialize)]
struct Artist {
    id: String,
    name: String,
    #[serde(rename = "type")]
    artist_type: Option<String>,
    country: Option<String>,
    #[serde(rename = "life-span")]
    life_span: Option<LifeSpan>,
    disambiguation: Option<String>,
    tags: Option<Vec<Tag>>,
}

#[derive(Debug, Deserialize)]
struct LifeSpan {
    begin: Option<String>,
    end: Option<String>,
    ended: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
struct Tag {
    name: String,
    count: i32,
}

#[derive(Debug, Deserialize)]
struct ReleaseSearchResponse {
    releases: Option<Vec<Release>>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Release {
    id: String,
    title: String,
    date: Option<String>,
    country: Option<String>,
    #[serde(rename = "artist-credit")]
    artist_credit: Option<Vec<ArtistCredit>>,
    #[serde(rename = "release-group")]
    release_group: Option<ReleaseGroup>,
}

#[derive(Debug, Deserialize)]
struct ArtistCredit {
    artist: ArtistRef,
}

#[derive(Debug, Deserialize)]
struct ArtistRef {
    name: String,
}

#[derive(Debug, Deserialize)]
struct ReleaseGroup {
    #[serde(rename = "primary-type")]
    primary_type: Option<String>,
}

impl MusicBrainzProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }

    fn search_artists(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/artist?query={}&limit={}&fmt=json",
            MUSICBRAINZ_API,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Ok(Vec::new());
        }

        let data: ArtistSearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .artists
            .unwrap_or_default()
            .into_iter()
            .map(|artist| {
                let lifespan = artist.life_span.as_ref();
                let years = match (lifespan.and_then(|l| l.begin.as_ref()), lifespan.and_then(|l| l.end.as_ref())) {
                    (Some(b), Some(e)) => format!(" ({} - {})", b, e),
                    (Some(b), None) => {
                        if lifespan.map(|l| l.ended.unwrap_or(false)).unwrap_or(false) {
                            format!(" ({} - ?)", b)
                        } else {
                            format!(" ({} - present)", b)
                        }
                    }
                    _ => String::new(),
                };

                let artist_type = artist.artist_type.as_deref().unwrap_or("Artist");
                let country = artist.country.as_ref().map(|c| format!(", {}", c)).unwrap_or_default();
                let disambiguation = artist.disambiguation.as_ref().map(|d| format!(" - {}", d)).unwrap_or_default();

                let top_tags: String = artist.tags
                    .as_ref()
                    .map(|tags| {
                        let mut sorted = tags.clone();
                        sorted.sort_by(|a, b| b.count.cmp(&a.count));
                        sorted.iter().take(3).map(|t| t.name.clone()).collect::<Vec<_>>().join(", ")
                    })
                    .filter(|s| !s.is_empty())
                    .map(|s| format!(". Genres: {}", s))
                    .unwrap_or_default();

                let mut metadata = HashMap::new();
                metadata.insert("type".to_string(), serde_json::json!("artist"));
                if let Some(t) = &artist.artist_type {
                    metadata.insert("artistType".to_string(), serde_json::json!(t));
                }
                if let Some(c) = &artist.country {
                    metadata.insert("country".to_string(), serde_json::json!(c));
                }
                if let Some(ls) = &artist.life_span {
                    if let Some(b) = &ls.begin {
                        metadata.insert("beginDate".to_string(), serde_json::json!(b));
                    }
                    if let Some(e) = &ls.end {
                        metadata.insert("endDate".to_string(), serde_json::json!(e));
                    }
                }

                KnowledgeEntry {
                    title: artist.name,
                    summary: format!("{}{}{}{}{}", artist_type, country, years, disambiguation, top_tags),
                    url: Some(format!("https://musicbrainz.org/artist/{}", artist.id)),
                    source: "musicbrainz".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }

    fn search_releases(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/release?query={}&limit={}&fmt=json",
            MUSICBRAINZ_API,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Ok(Vec::new());
        }

        let data: ReleaseSearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .releases
            .unwrap_or_default()
            .into_iter()
            .map(|release| {
                let artists = release.artist_credit
                    .as_ref()
                    .map(|ac| ac.iter().map(|a| a.artist.name.clone()).collect::<Vec<_>>().join(", "))
                    .unwrap_or_else(|| "Unknown artist".to_string());

                let year = release.date
                    .as_ref()
                    .map(|d| format!(" ({})", d.get(..4).unwrap_or(d)))
                    .unwrap_or_default();

                let release_type = release.release_group
                    .as_ref()
                    .and_then(|rg| rg.primary_type.as_ref())
                    .map(|t| t.as_str())
                    .unwrap_or("Release");

                let mut metadata = HashMap::new();
                metadata.insert("type".to_string(), serde_json::json!("release"));
                if let Some(rg) = &release.release_group {
                    if let Some(pt) = &rg.primary_type {
                        metadata.insert("releaseType".to_string(), serde_json::json!(pt));
                    }
                }
                if let Some(d) = &release.date {
                    metadata.insert("date".to_string(), serde_json::json!(d));
                }

                KnowledgeEntry {
                    title: release.title,
                    summary: format!("{} by {}{}", release_type, artists, year),
                    url: Some(format!("https://musicbrainz.org/release/{}", release.id)),
                    source: "musicbrainz".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }
}

impl Default for MusicBrainzProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for MusicBrainzProvider {
    fn name(&self) -> &'static str {
        "musicbrainz"
    }

    fn is_available(&self) -> bool {
        self.client
            .get(format!("{}/artist?query=test&limit=1&fmt=json", MUSICBRAINZ_API))
            .send()
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        let mut entries = match self.search_artists(query, limit) {
            Ok(e) => e,
            Err(e) => return LookupResult::error(self.name(), e),
        };

        if entries.len() < limit {
            let remaining = limit - entries.len();
            match self.search_releases(query, remaining) {
                Ok(releases) => entries.extend(releases),
                Err(e) => return LookupResult::error(self.name(), e),
            }
        }

        LookupResult::success(self.name(), entries)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn musicbrainz_lookup() {
        let provider = MusicBrainzProvider::new();
        let result = provider.lookup("Beatles", &LookupOptions::default());
        assert!(result.success);
        assert!(!result.entries.is_empty());
    }
}
